### PR TITLE
Add Vagrantfile to provide platform independent development environment

### PR DESCRIPTION
**Pull request by t-mat**
This PR adds Vagrantfile and its provisioning script which provides Linux VM for development of xxhash.  This VM will run on Linux, MacOSX and Windows.

Basic goal of this PR is providing platform independent development environment for major platforms to test/develop xxHash.  Even on Linux, since VM provides clean environment, it would be useful to test environment dependent problems.

You can check this PR by the following procedure
**NOTE : This procedure takes a long time.  Maybe 5 to 30 minutes**
### Prerequisite

This procedure needs the following softwares:
- [git](https://git-scm.com/)
- [VirtualBox](https://www.virtualbox.org/)
- [Vagrant](https://www.vagrantup.com/)
### Setup on Linux

```
cd
git clone https://github.com/t-mat/xxHash.git xxhash-tmat
cd xxhash-tmat
git checkout virtual-devenv
cd devenv/linux

vagrant up
vagrant ssh

## If you see passphrase prompt, just press Enter ##
## If you see password prompt for vagrant, input "vagrant" as password ##
## here, you'll see GUEST terminal ##
# note: GUEST's ~/xxhash/ is mapped to HOST's ~/xxhash-tmat/devenv/linux/xxhash/ #
cd xxhash
git checkout dev
make all
./xxhsum --help
make test-all
exit
## here, you'll back to the HOST ##

vagrant halt     # halt VM #
vagrant status   # show VM status #

# If you don't need VM anymore, destroy VM and delete shared directory #

vagrant destroy  -f # delete VM image, if you want #
rm -rf xxhash       # delete cloned repository #
```
### Possible problems
- It uses old [Ubuntu Precise (12.04LTS, 64bit)](http://releases.ubuntu.com/12.04/).
  - I'm not sure newer version is good for testing/development.
  - Since there are many Vargrant VM images, we can change it easily.
- I'm not sure which is better for cloned repository name in lowercase `xxhash` or camelCase `xxHash`.

### CODE REVIEW

**Comment by Cyan4973** [comment]
Hi @t-mat 

I initially incorrectly understood your patch would make it possible to test multiple target systems (arm, powerpc, mips, etc.), like Qemu user mode, but this is different. If I do understand, the idea is to ensure 2 developers environment can recreate identical dev conditions. In principle, this is a good idea.

So, I've been testing vagrant and followed your instructions manually on my own linux dev station.
Unfortunately, the process proved a little bit harder than expected.

Installing vagrant proved not enough. I also had to install a few other dependencies (don't even remembered which one) that `apt-get` wouldn't know about. Discovering them one by one was a manual process of try, error, analysis, install.
After finally getting no more unsolved dependencies, `vagrant up` would download something (which indeed takes some time) but ultimately fail to launch for a few different reasons, which were also solved one by one.
Last one was : `The machine is in the 'poweroff' state. Please verify everything is configured properly and try again.`

at which stage I stopped.

Even if this issue can also be solved, this seems way too many difficulties to get it running.

Basically, vagrant selling point is to make dev systems free variability : " Say goodbye to the "works on my machine" excuse ". Apparently, my experience is that vagrant replaced a bunch of dependencies with another set of dependencies, now centered on vagrant running correctly. And I haven't even tested Windows and Mac OS X, which could be different.

I can understand that my experience might have been inferior to others. I can also understand that some development environments require a huge amount of carefully built dependencies, which are difficult to recreate. In which case, such project would hugely benefit from such an approach.

But xxhash is much simpler, and its dependencies are pretty much limited to standard C libs and runtime environment. Does vagrant help then ?

**Comment by Cyan4973** [comment]
Btw, there are so many levels of virtualization that I can get lost sometimes.

Is it possible to explain how Vagrant differs from something like Docker ?

**Comment by t-mat** [comment]
Sorry for your bad experience :fearful:  I didn't want to introduce another level of complexity, but it seems I'm failed.  Please feel free to reject this PR.

Since this Vagrantfile uses most basic Vagrant box [(`hashicorp/precise64`) which has been used in the first tutorial of Vagrant](https://www.vagrantup.com/docs/getting-started/) and it've been tested by many tutorial users, I'm surprised that you met difficulty.  
It seems my Vagrantfile has problem.  Could you tell me your current Linux distro and version?  I'd like to investigate it.

> Does vagrant help then ?

Frankly, Linux-VM on Linux host doesn't help much.
But if I'll solve your problem and this PR will be accepted successfully, I would like to expand this approach for another VM such as FreeBSD-VM and Windows(Server)-VM on Linux host.  I hope these VM (sometimes) will help actual development and testing.

> Is it possible to explain how Vagrant differs from something like Docker ?

From user's point of view, basically Docker is filesystem/network sandbox for Linux.  All Docker image runs on host's Linux kernel but access different filesystem/network.  In other words, Docker doesn't emulate another Machine/CPU/OS.  It's always (userspace) Linux-VM on Linux host.

On the other hands, Vagrant image runs on actual virtual machine emulator.  Vagrant always emulates another Machine/CPU/OS on VM.  (Technically, Vagrant just invokes 3rd party emulator such as VirtualBox or qemu.)  Even Linux-VM on Linux host, Linux-VM runs on virtual machine.

First 50 lines of "[A Not Very Short Introduction to Docker](http://www.jayway.com/2015/03/21/a-not-very-short-introduction-to-docker/)" represents difference between Docker and Vagrant well.

**Comment by Cyan4973** [comment]
> Could you tell me your current Linux distro and version?

I tested it on Linux Mint 17.1

>  I would like to expand this approach for another VM such as FreeBSD-VM and Windows(Server)-VM on Linux host. 

This sounds like an excellent perspective

> Vagrant always emulates another Machine/CPU/OS on VM. (Technically, Vagrant just invokes 3rd party emulator such as VirtualBox or qemu.) 

This looks interesting. So, it would be possible to combine it with Qemu in order to test non-x86 CPU (ARM, MIPS, PowerPC, etc.), typically on Travis CI ?

> NOTE : This procedure takes a long time. Maybe 5 to 30 minutes

Well, this could be embarrassing if the objective is to run tests automatically on Travis CI.
Time out could get in the way.

That's maybe why I've read people using Docker to emulate different systems. It could be a question of startup speed.

Thanks to your explanation, the technical differences between Vagrant and Docker look much clearer.

If the point of Vagrant is to run emulated target, it should be more adapted for this task than loading a Docker conf. in which there is a prepared emulator to do the same thing.

In a nutshell, I'm all for it. This is a great objective.

Now let's ensure installation experience get a little better...

#### Context

I've been trying to run tests on different targets using Qemu user mode on Travis CI.
Considering the accessible content from Travis VM (using `apt-get`), only ARM target works.
That's pretty limited, especially as emulated ARM is little endian, and one of the objectives is to test big-endian CPU.

On top of that, it's obviously limited to Linux.

If something more generic like Vagrant can be made to work, this would be a great plus for testing abilities.

**Comment by Cyan4973** [comment]
I made some new tests with `virtual-devenv` branch.

This time, instead of following manually the tutorial, I would just go the prepared branch and run the script. Here are some results : 

```
./setup.sh 
-- begin : xxhash provisioning --
./setup.sh: 18: ./setup.sh: cannot create /etc/apt/sources.list.d/0.list: Permission denied
./setup.sh: 19: ./setup.sh: cannot create /etc/apt/sources.list.d/0.list: Permission denied
./setup.sh: 20: ./setup.sh: cannot create /etc/apt/sources.list.d/0.list: Permission denied
./setup.sh: 21: ./setup.sh: cannot create /etc/apt/sources.list.d/0.list: Permission denied
./setup.sh: 22: ./setup.sh: cannot create /etc/apt/sources.list.d/0.list: Permission denied
Password: 
W: GPG error : http://download.opensuse.org  Release : Unverified signature due to missing public key  : NO_PUBKEY DA4C453AE9417287
Sélection g++-multilib 
Préparation du décompactage de .../g++-multilib_4%3a4.8.2-1ubuntu6_amd64.deb ...
Décompactage de g++-multilib (4:4.8.2-1ubuntu6) ...
Paramétrage de g++-multilib (4:4.8.2-1ubuntu6) ...
fatal: cannot create directories into '/vagrant/xxhash': no Permission
-- end : xxhash provisioning --
```

The script prompted me for the sudo password, which I correctly entered. Nonetheless, a few operations failed, due to missing administrator rights.

These problems were mostly solved using : 

```
sudo ./setup.sh 

-- begin : xxhash provisioning --
W: GPG error : http://download.opensuse.org  Release : Unverified signature due to missing public key  : NO_PUBKEY DA4C453AE9417287
Cloning into '/vagrant/xxhash'...
remote: Counting objects: 558, done.
remote: Compressing objects: 100% (51/51), done.
Réception d'objets: 100% (558/558), 432.10 KiB | 701.00 KiB/s, done.
Résolution des deltas: 100% (305/305), done.
remote: Total 558 (delta 28), reused 0 (delta 0), pack-reused 507
Vérification de la connectivité... fait.
-- end : xxhash provisioning --
```

Only the GPG verification warning remains, which isn't a blocking issue.

Now onto next step : 

```
vagrant up
Bringing machine 'default' up with 'virtualbox' provider...
There are errors in the configuration of this machine. Please fix
the following errors and try again:

vm:
* The box 'hashicorp/precise64' could not be found.
```

So likely an installation problem.

As said previously, vagrant and dependencies installations were a pain on my system, so something could have gone wrong.

Alternatively, maybe it could be a good idea for `setup` to also ensure proper installation of vagrant and dependencies (such as VirtualBox).
After all, there is already a section installing a bunch of things.


###### Quick question : 
What is `/etc/apt/sources.list.d/0.list` ?

**Comment by t-mat** [comment]
> I tested it on Linux Mint 17.1

Thanks.  I'll investigate it.

> So, it would be possible to combine it with Qemu in order to test non-x86 CPU (ARM, MIPS, PowerPC, etc.), typically on Travis CI ?

I've never tried to use Vagrant for non-x86/64 architecture, but it seems there is possibility.

 - https://github.com/pradels/vagrant-libvirt/issues/286
 - http://cloudgeekz.com/254/how-to-use-vagrant-with-powerkvm.html

There are also interesting posts in Travis-CI's issue:

 - [Support Android / Arm](https://github.com/travis-ci/travis-ci/issues/3376)
    - https://github.com/travis-ci/travis-ci/issues/3376#issuecomment-158759546
    - https://github.com/travis-ci/travis-ci/issues/3376#issuecomment-168003756

It seems [multiarch/qemu-user-static](multiarch/qemu-user-static) is suitable for our testing purpose.  It's Docker image which contains usual [`qemu-*-static` executable files](https://github.com/multiarch/qemu-user-static/releases/).  I don't test these Docker images on TravisCI.  But if it works fine, it's good alternative.

> This time, instead of following manually the tutorial, I would just go the prepared branch and run the script. 

Sorry for bad naming!  `setup.sh` is special shell script for Vagrant's provisioning.  You can't use it as usual shell script.  If `vagrant up` works properly, `setup.sh` will be running in Vagrant's **guest** VM.

> What is /etc/apt/sources.list.d/0.list ?

In short: It's mirror server list to select best (nearest in network distance) server.  But I think my method was bad, I must change it.

Tweaking for `/etc/apt/sources.list` and `/etc/apt/sources.list.d/0.list` is quick hack to improve `apt-get update` speed.  Since default `hashicorp/precise64` image's `/etc/apt/sources.list` (server list for `apt-get`) always look up for US mirror server, `apt-get update` becomes quite slow when host machine is outside of US.
I've copied mirror server list from one of answer for ["How do you select the fastest mirror from the command line?"](http://askubuntu.com/a/9035).

### todo

 - [ ] Investigate Vagrant setup problem
 - [ ] Check existence of actual working Vagrantfile for non-x86/64 architecture on Linux host.
 - [ ] Try multiarch/qemu-user-static
 - [ ] Try TravisCI + multiarch/qemu-user-static

**Comment by Cyan4973** [comment]
:+1: 

Quick question : with the integration of #56 , I was considering the release of a new version of xxHash.  

Should I wait for #57 ?

**Comment by Cyan4973** [comment]
@t-mat : is this topic still opened ?
